### PR TITLE
MAP: change bool to float32

### LINKED ISSUE #1147: RuntimeError when using MAP-metric
## 🐛 Bug

Hi! I am training a detection model and use MAP-metric during validation.  I got the following error at the validation_step: 
`RuntimeError: expected scalar type Float but found Bool`.

### To Reproduce

Pick a faster rcnn model, I used `fasterrcnn_resnet50_fpn_v2`() from `torchvision` . Implement validation_step where
`self.metrics.update(...)` is called for the model results and targets and validation_epoch_end where the `self.metrics.compute()` is called for the previously gathered results. 

#### Code sample

```py
import torchvision
from torchvision.models.detection.faster_rcnn import FastRCNNPredictor
from torchmetrics.detection.mean_ap import MeanAveragePrecision


class FasterRCNNModel(pl.LightningModule):
    def __init__(self, num_classes):
        super().__init__()

        model = torchvision.models.detection.faster_rcnn.fasterrcnn_resnet50_fpn_v2()
        in_features = model.roi_heads.box_predictor.cls_score.in_features
        model.roi_heads.box_predictor = FastRCNNPredictor(in_features, num_classes)
        self.model = model
        self.metric = MeanAveragePrecision(box_format='xyxy', iou_type='bbox')

    def validation_step(self, batch, batch_idx):
        images, targets = batch
        preds = self.model(images)        
        self.metric.update(preds, targets)

    def validation_epoch_end(self, outs):
        mAP = self.metric.compute()
        self.log("val/mAP", mAP)
        self.metric.reset()
```
targets (List[Dict]), containing:
- boxes (`torch.float32`)
- labels (`torch.int64`)

preds (List[Dict]), containing:
- boxes (`torch.float32`)
- scores (`torch.float32`)
- labels (`torch.int64`) 

#### Error message
```
  File "/homes/vsoboleva/scripts/pascal_voc/train.py", line 65, in validation_epoch_end
    mAP = self.metric.compute()
  File "/homes/vsoboleva/miniconda3/lib/python3.9/site-packages/torchmetrics/metric.py", line 523, in wrapped_func
    value = compute(*args, **kwargs)
  File "/homes/vsoboleva/miniconda3/lib/python3.9/site-packages/torchmetrics/detection/mean_ap.py", line 908, in compute
    precisions, recalls = self._calculate(classes)
  File "/homes/vsoboleva/miniconda3/lib/python3.9/site-packages/torchmetrics/detection/mean_ap.py", line 758, in _calculate
    recall, precision, scores = MeanAveragePrecision.__calculate_recall_precision_scores(
  File "/homes/vsoboleva/miniconda3/lib/python3.9/site-packages/torchmetrics/detection/mean_ap.py", line 831, in __calculate_recall_precision_scores
    det_scores = torch.cat([e["dtScores"][:max_det] for e in img_eval_cls_bbox])
RuntimeError: expected scalar type Float but found Bool
```

### Expected behavior

The `sel.metric.compute(...)` compute values correctly and does not fail with `RuntimeError: expected scalar type Float but found Bool`.

### Environment

- TorchMetrics 0.9.2 build with pip
- Python 3.9.12, torch 1.12.0, torchvision 0.13.0
- OS (e.g., Linux): Ubuntu 20.04.3

### PR DESCRIPTION
**Pull request by dreaquil**
## What does this PR do?

Fixes #1147 

## Before submitting

- [x] Was this **discussed/approved** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/metrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to **update the docs**?
- [x] Did you write any new **necessary tests**?

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: src/torchmetrics/detection/mean_ap.py [hunk 1]
```diff
@@ -502,7 +502,7 @@ def __evaluate_image_gt_no_preds(
         return {
             "dtMatches": torch.zeros((nb_iou_thrs, nb_det), dtype=torch.bool, device=self.device),
             "gtMatches": torch.zeros((nb_iou_thrs, nb_gt), dtype=torch.bool, device=self.device),
-            "dtScores": torch.zeros(nb_det, dtype=torch.bool, device=self.device),
+            "dtScores": torch.zeros(nb_det, dtype=torch.float32, device=self.device),
```
Comment (Borda) [inline src/torchmetrics/detection/mean_ap.py:505]: can we pls add a test to cover this case?

Comment (dreaquil) [reply src/torchmetrics/detection/mean_ap.py:505]: @SkafteNicki I am currently unable to retrieve an input that causes the error but I think the change is fairly minute and straightforward. dtScores should be a tensor of type float32 according to documentation.

**Comment by ashnair1** [comment]
Any update on this PR?

**Review by justusschock — APPROVED** [review]

File: src/torchmetrics/detection/mean_ap.py (hunk 1 above)
Comment (SkafteNicki) [reply src/torchmetrics/detection/mean_ap.py:503]: test case added from this comment: https://github.com/Lightning-AI/metrics/issues/1147#issuecomment-1198910780

**Review by SkafteNicki — APPROVED** [review]

**Pull request merged**